app/models/mixins/supports_feature_mixin.rb

```diff
@@ -226,35 +119,45 @@ def unsupported
 
   class_methods do
     # This is the DSL used a class level to define what is supported
-    def supports(feature, &block)
-      SupportsFeatureMixin.guard_queryable_feature(feature)
-      define_supports_feature_methods(feature, &block)
+    def supports(feature, description: nil, &block)
+      define_supports_feature_methods(feature, :description => description, &block)
     end
 
     # supports_not does not take a block, because its never supported
     # and not conditionally supported
-    def supports_not(feature, reason: nil)
-      SupportsFeatureMixin.guard_queryable_feature(feature)
-      define_supports_feature_methods(feature, :is_supported => false, :reason => reason)
+    def supports_not(feature, description: nil, reason: nil)
+      define_supports_feature_methods(feature, :is_supported => false, :description => description, :reason => reason)
     end
 
     # query the class if the feature is supported or not
     def supports?(feature)
-      SupportsFeatureMixin.guard_queryable_feature(feature)
+      guard_queryable_feature(feature)
       public_send("supports_#{feature}?")
     end
 
     # query the class for the reason why something is unsupported
     def unsupported_reason(feature)
-      SupportsFeatureMixin.guard_queryable_feature(feature)
+      guard_queryable_feature(feature)
       feature = feature.to_sym
       public_send("supports_#{feature}?") unless unsupported.key?(feature)
       unsupported[feature]
     end
 
     # query the class if a feature is generally known
     def feature_known?(feature)
-      SupportsFeatureMixin::QUERYABLE_FEATURES.key?(feature.to_sym)
+      queryable_features.key?(feature.to_sym)
+    end
+
+    def guard_queryable_feature(feature)
+      unless queryable_features.key?(feature.to_sym)
+        raise UnknownFeatureError, "Feature ':#{feature}' is unknown to SupportsFeatureMixin."
+      end
+    end
+
+    def queryable_features
+      # This is a class variable and it might be modified during runtime
+      # because we do not eager load all classes at boot time, so it needs to be thread safe
+      @queryable_features ||= Concurrent::Hash.new
```

Comment:
`||=` is not thread safe either (i.e. 2 threads could create 2 separate concurrent hashes), but I'm not sure that matters?  @jrafanie Thoughts here?

Reply:
You can use something  like `AtomicReference#compare_and_set(nil, Concurrent::Hash.new
@Fryguy )`

Reply:
I just copied what was done for the `unsupported` method.  If we change one, we need to change both.  Maybe use a Mutex somehow instead?